andela/wolfsbane-backend
# #167750057 Implement user profile page functionality

**What does this PR do?**
This PR implements user profile page functionality

**Description of tasks to be completed?**
- write tests to test profile page endpoints
- write routes to get, create, and update profile details
- write middleware and controllers to handle the routes
- test enpoints functionality using postman

**How should this be manually tested/checked?**
1. In your terminal, run `git clone https://github.com/andela/wolfsbane-backend.git`
2. Run `cd wolfsbane-backend`
3. Run `git fetch origin ft-user-profile-settings-#167750057`
4. Run `npm install`
5. Run `npm run migrate` to migrate the model definitions to both databases (`development` and `test`)
6. Run `npm run seed` to seed the database with the needed data
7. Run `npm run start:dev` to start the app
8. In postman, set the url to`http://localhost:3000/api/v1/users/signup`, set method to `POST` and then send a JSON request object like this:
```json
{
    "email" : "[Enter your email here]",
    "password" : "[Enter a good password]",
    "firstName" : "[Enter your first name]",
    "lastName" : "[Enter your last name]"
}
```
9. Check your email and click on the verification link sent to you.
10. In postman, set the url to`http://localhost:3000/api/v1/users/signin`, set method to `POST` and then send a JSON request object like this:
```json
{
    "email" : "[The email you registered with]",
    "password" : "[The password you registered with]",
}
```
11. Copy the token property in the response object
12. Click the `Authorization` tab and in the `Type` dropdown, choose `Bearer Token`, then paste the `token` you copied into the token input field.
13. Then set the url to`http://localhost:3000/api/v1/profiles`, and the method to `POST`, and send the following details:
```json
{
    "phoneNumber" : "222-333-444",
    "department" : "Software Engineering",
    "jobDescription" : "Software Developer",
    "imageUrl" : "myImage.png"
}
```
14. You should get a success message informing you that profile was created successfully, alongside the profile details that looks like this:
```json
{
    "status": 200,
    "message": "Profile created successfully",
    "data": {
        "id": "763ff831-d135-4aa8-b0f2-e7d2bbd84663",
        "userId": "5041a1d7-cc37-47e3-8c2d-4031331dbf4d",
        "phoneNumber": "222-333-444",
        "role": "Staff",
        "jobDescription": "Software Developer",
        "imageUrl": "myPicture.png",
        "theDepartment": {
            "department": "Software Engineering"
        },
        "theUser": {
            "firstName": "John",
            "lastName": "Doe",
            "email": "johndoe@gmail.com",
            "isVerified": true
        }
    }
}
```
15. Change the request method to `GET` , copy the `userId` property of the response you got from the just created profile, add it to the url in the url tab to look something like `http:localhost:3000/api/v1/profiles/:userId` and send. Your response should look like this:
```json
{
    "status": 200,
    "message": "Profile fetched successfully",
    "data": {
        "phoneNumber": "222-333-444",
        "role": "Staff",
        "jobDescription": "Software Developer",
        "imageUrl": "myPicture.png",
        "userId": "5041a1d7-cc37-47e3-8c2d-4031331dbf4d",
        "theDepartment": {
            "department": "Software Engineering"
        },
        "theUser": {
            "firstName": "John",
            "lastName": "Doe",
            "email": "johndoe@gmail.com",
            "isVerified": true
        }
    }
}
```
16.Change the request method to `PUT` , edit any detail in the json data that was used to create the profile and send. Your response should look like this:
```json
{
    "status": 200,
    "message": "Profile updated successfully",
    "data": {
        "id": "763ff831-d135-4aa8-b0f2-e7d2bbd84663",
        "userId": "5041a1d7-cc37-47e3-8c2d-4031331dbf4d",
        "phoneNumber": "222-444-555",
        "role": "Staff",
        "jobDescription": "Technical Team Lead",
        "imageUrl": "myNewPicture.png",
    }
}
```
17.To test the `delete` functionality, open a new tab and set the address to `http://localhost:3000/api/v1/users/signin` and sign in with the super admin details in the `background details` below. Copy the token returned and put it in the token tab of the former request tab. Then change the request method to `DELETE` and send. Your response should look like this:
```json
{
    "status": 200,
    "message": "Profile deleted successfully",
}
```
18.To test the `change roles` make sure you are logged in as the super admin and have copied your token to the Authorization tab. Set your address to `http://localhost:3000/api/v1/profiles/:userId/roles`, and the method to `PUT`. Send a `json` request like this:
```json
{
    "role": "Manager"
}
```
You should get a `response`
```json
{
    "status": 200,
    "message": "Profile role updated successfully",
}
```


**Any background context you want to provide?**
- You need to input a valid email address in the `signup` request so you can get the mail to verify your account, as you won't be granted any access to the app until you do so.
- Note that to test super admin routes, like `delete`, and `role update`, you have to be signed in as a super admin. The login details are:
```json
{
       "email": "funmi1@gmail.com",
       "password": "funmi1234"
}
```

**What are the relevant pivotal tracker stories?**
[#167750057](https://www.pivotaltracker.com/story/show/167750057)
